iup/__init__.py

```diff
@@ -0,0 +1,358 @@
+                .sort("date")
+                .filter(pl.col("date") >= start_date)
+            )
+
+        out = IncidentUptakeData(out)
+        out.validate()
+
+        return out
+
+    def trim_outlier_intervals(self) -> Self:
+        self = self.sort("date").with_row_count("index")
+        unique_regions = self["region"].unique()
+        sub_frames = []
+
+        for x in range(len(unique_regions)):
+            this_region = pl.col("region") == unique_regions[x]
+            if (standardize(self.filter(this_region)["interval"]))[1] > 1:
+                sub_frames.append(self.filter(this_region).slice(3))
+            else:
+                sub_frames.append(self.filter(this_region).slice(2))
+
+        self = pl.concat(sub_frames).sort("date")
+
+        return self
+
+    def to_cumulative(self, last_cumulative):
+        self = (
+            self.select(["region", "date", "estimate"])
+            .sort("date")
+            .with_columns(estimate=pl.col("estimate").cum_sum().over("region"))
+            .join(last_cumulative, on="region")
+            .with_columns(estimate=pl.col("estimate") + pl.col("last_cumulative"))
+            .drop("last_cumulative")
+        )
+
+        self = CumulativeUptakeData(self)
+
+        return self
+
+
+class CumulativeUptakeData(UptakeData):
+    def validate(self):
+        self.assert_columns_found(["region", "date", "estimate"])
+        self.assert_columns_type(["date"], pl.Date)
+        self.assert_columns_type(["region"], pl.Utf8)
+        self.assert_columns_type(["estimate"], pl.Float64)
+        assert (
+            self["estimate"] >= 0.0
+        ).all(), "Cumulative uptake estimates cannot be negative."
+
+    def to_incident(self) -> IncidentUptakeData:
+        self = (
+            self.with_columns(
+                season=(
+                    pl.col("date").dt.year()
+                    + pl.when(pl.col("date").dt.month() < 7).then(-1).otherwise(0)
+                ).cast(pl.Utf8),
+                estimate=pl.col("estimate").diff().over("region"),
+                elapsed=(pl.col("date") - pl.col("date").first())
+                .over("region")
+                .dt.total_days()
+                .cast(pl.Float64),
+            )
+            .with_columns(
+                estimate=pl.col("estimate")
+                .fill_null(pl.col("estimate").first())
+                .over("region"),
+                interval=pl.col("elapsed").diff().over("region"),
+            )
+            .with_columns(
+                interval=pl.col("interval")
+                .fill_null(pl.col("elapsed").first())
+                .over("region")
+            )
+            .with_columns(estimate=(pl.col("estimate") / pl.col("interval")))
+            .with_columns(previous=pl.col("estimate").shift(1).over("region"))
+        )
+
+        self = IncidentUptakeData(self)
+
+        return self
+
+    @staticmethod
+    def split_train_test(uptake_data_list, start_date, side):
```

Comment:
This seems identical to the other function, above? That tells me it should be part of the parent class?